src/nb_func.cpp

```diff
@@ -432,8 +432,31 @@ PyObject *nb_func_new(const void *in_) noexcept {
         type_data *td = nb_type_data((PyTypeObject *) f->scope);
         bool has_new = td->flags & (uint32_t) type_flags::has_new;
 
-        if (is_init && !has_new) {
-            td->init = func;
+        if (is_init) {
+            if (!has_new) {
+                td->init = func;
+            } else {
+                // Keep track of whether we have a __init__ overload that
+                // accepts no arguments (except self). If not, then we
+                // shouldn't allow calling the type object with no arguments,
+                // even though (for unpickling support) we probably do have
+                // a __new__ overload that accepts no arguments (except cls).
+                // This check is necessary because our type vectorcall shortcut
+                // skips Python's usual logic where __init__ is always called
+                // if __new__ returns an instance of the type.
+                bool noargs_ok = true;
+                for (size_t i = 1; i < fc->nargs - has_var_kwargs; ++i) {
```

Comment:
Could the loop here also use `i < fc->nargs - has_var_kwargs - has_var_args` as stopping criterion?

Reply:
No, because there can be keyword-only parameters in between `nb::args` and `nb::kwargs`, and the function is still invocable with no arguments as long as they all have defaults.

Reply:
Right, thank you!